pkg/pillar/hypervisor/kubevirt.go

```diff
@@ -0,0 +1,1421 @@
+				}
+				parsedValue = value
+			} else {
+				// Try parsing as int
+				value, err := strconv.Atoi(metricValue)
+				if err != nil {
+					continue
+				}
+				parsedValue = value
+			}
+
+			var domainName string
+			for n, vmis := range ctx.vmiList {
+				if vmis.name == vmiName {
+					domainName = n
+					if _, ok := res[domainName]; !ok {
+						res[domainName] = types.DomainMetric{
+							UUIDandVersion: types.UUIDandVersion{},
+							CPUScaled:      1,
+						}
+					}
+				}
+			}
+			fillMetrics(res, domainName, metricName, parsedValue)
+			logrus.Debugf("GetDomsCPUMem: vmi %s, domainName %s, metric name %s, value %v", vmiName, domainName, metricName, parsedValue)
+		}
+	}
+
+	for n, r := range res {
+		if n == "" {
+			continue
+		}
+		// used_bytes = available_bytes - usable_bytes
+		r.UsedMemory = r.UsedMemory - r.AvailableMemory
+		if r.AllocatedMB > 0 {
+			per := float64(r.UsedMemory) / float64(r.AllocatedMB)
+			r.UsedMemoryPercent = per
+		}
+		if r.UsedMemory > r.MaxUsedMemory {
+			r.MaxUsedMemory = r.UsedMemory
+		}
+		if _, ok := ctx.prevDomainMetric[n]; !ok {
+			r.MaxUsedMemory = r.UsedMemory
+			ctx.prevDomainMetric[n] = r
+		} else {
+			if ctx.prevDomainMetric[n].MaxUsedMemory > r.UsedMemory {
+				r.MaxUsedMemory = ctx.prevDomainMetric[n].MaxUsedMemory
+			} else if ctx.prevDomainMetric[n].MaxUsedMemory < r.UsedMemory {
+				r.MaxUsedMemory = r.UsedMemory
+			}
+			ctx.prevDomainMetric[n] = r
+		}
+		res[n] = r
+	}
+
+	hasEmptyRes := len(ctx.vmiList) - len(res)
+	if hasEmptyRes > 0 {
+		// check and get the kubernetes pod's metrics
+		checkPodMetrics(ctx, res, hasEmptyRes)
+	}
+	logrus.Debugf("GetDomsCPUMem: %d VMs: %+v, podnum %d", len(ctx.vmiList), res, hasEmptyRes)
+	return res, nil
+}
+
+func getVirtHandlerIPAddr(ctx *kubevirtContext) (string, error) {
+	if ctx.virthandlerIPAddr != "" {
+		return ctx.virthandlerIPAddr, nil
+	}
+	clientSet, err := kubeapi.GetClientSet()
+	if err != nil {
+		return "", err
+	}
+
+	pods, err := clientSet.CoreV1().Pods("kubevirt").List(context.Background(),
+		metav1.ListOptions{
+			LabelSelector: "kubevirt.io=virt-handler",
+		})
+	if err != nil {
+		return "", err
+	}
+
+	var vmiPod *k8sv1.Pod
+	for _, pod := range pods.Items {
+		if strings.HasPrefix(pod.ObjectMeta.Name, "virt-handler-") {
+			vmiPod = &pod
+			break
+		}
+	}
+
+	if vmiPod == nil {
+		return "", fmt.Errorf("can not find virt-handler pod")
+	}
+	ctx.virthandlerIPAddr = vmiPod.Status.PodIP
+	return ctx.virthandlerIPAddr, nil
+}
+
+func getVMIName(metricStr string) string {
+	nameStr := strings.SplitN(metricStr, ",name=\"", 2)
+	if len(nameStr) != 2 {
+		fmt.Printf("get name failed, string %s\n", metricStr)
+		return ""
+	}
+	nameStr2 := strings.SplitN(nameStr[1], "\",", 2)
+	if len(nameStr2) != 2 {
+		fmt.Printf("get name2 failed, string %s\n", metricStr)
+		return ""
+	}
+	return nameStr2[0]
+}
+
+func fillMetrics(res map[string]types.DomainMetric, domainName, metricName string, value interface{}) {
```

Comment:
but this only passes in the reference to the map of res as far as I know of, right?

Reply:
Yes, but if it is not clear, then it would perhaps be good to have a test for it ;-)

Still, the questions stays, why is it needed to pass the whole map?

Reply:
I think in this function, we update the results to the map w/ this index, so the function handles itself and no need to return value and search for the entry in the map to update. Or do i miss something here?

Reply:
`fillMetrics` (plural) sounds like this function would fill more than one metric, and then by passing the whole map it again suggests that several fields will be updated.
But it does not.

Why not just do something like:
```
type domainMetrics map[string]types.DomainMetric

func (metrics *domainMetrics) fill(domainName, metricName string, value interface{}) {
```

Reply:
sure, @zedi-pramodh i have a patch with this:
[vmi-metrics-diff.txt](https://github.com/lf-edge/eve/files/15067714/vmi-metrics-diff.txt)

Reply:
ack, will be in next commit